File: facultebackend/src/main/java/tn/faculte/facultebackend/Service/ReportService.java
package tn.faculte.facultebackend.Service;

import org.springframework.http.ResponseEntity;
import tn.faculte.facultebackend.Dto.CommentDto;
import tn.faculte.facultebackend.Dto.PostDto;
import tn.faculte.facultebackend.Dto.ReportedCommentDto;
import tn.faculte.facultebackend.Dto.ReportedPostDto;
import tn.faculte.facultebackend.Entity.Comment;
import tn.faculte.facultebackend.Entity.Post;
import tn.faculte.facultebackend.Entity.Report;

import java.util.List;

public interface ReportService {
    Report getReportById(Long id);

    Report reportPost(Post post, String reason);
    Report reportComment(Comment comment, String reason);
    List<Report> getAllReports();
    List<Report> getAllPendingReports();
    void deleteReport(Report report);
    void deleteReportedPost(Report report);
    void deleteReportedComment(Report report);
    public List<ReportedCommentDto> getReportedComments();
    public List<ReportedPostDto> getReportedPosts();
}

File: Faculte.tn/src/app/Services/report.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

@Injectable({
  providedIn: 'root'
})
export class ReportService {

  private apiUrl = 'https://facultetnbackend.onrender.com/api/reports';

  constructor(private http: HttpClient) { }

  private getHeaders(): HttpHeaders {
    const token = localStorage.getItem('accessToken');
    if (token) {
      return new HttpHeaders({
        'Authorization': `Bearer ${token}`
      });
    } else {
      // Handle the case where token is not available
      return new HttpHeaders();
    }
  }

  reportPost(postId: number, reason: string): Observable<any> {
    const headers = this.getHeaders();
    return this.http.post<any>(`${this.apiUrl}/reportPost?postId=${postId}&reason=${reason}`, null, { headers });
  }

  reportComment(commentId: number, reason: string): Observable<any> {
    const headers = this.getHeaders();
    return this.http.post<any>(`${this.apiUrl}/reportComment?commentId=${commentId}&reason=${reason}`, null, { headers });
  }

  getAllReports(): Observable<any[]> {
    const headers = this.getHeaders();
    return this.http.get<any[]>(`${this.apiUrl}`, { headers });
  }

  getAllPendingReports(): Observable<any[]> {
    const headers = this.getHeaders();
    return this.http.get<any[]>(`${this.apiUrl}/pending`, { headers });
  }

  deleteReport(id: number): Observable<void> {
    const headers = this.getHeaders();
    return this.http.delete<void>(`${this.apiUrl}/${id}`, { headers });
  }

  deleteReportedPost(id: number): Observable<void> {
    const headers = this.getHeaders();
    return this.http.delete<void>(`${this.apiUrl}/deleteReportedPost/${id}`, { headers });
  }

  deleteReportedComment(id: number): Observable<void> {
    const headers = this.getHeaders();
    return this.http.delete<void>(`${this.apiUrl}/deleteReportedComment/${id}`, { headers });
  }
  getReportedPosts(): Observable<any[]> {
    const headers = this.getHeaders();
    return this.http.get<any[]>(`${this.apiUrl}/reportedPosts`, { headers });
  }

  getReportedComments(): Observable<any[]> {
    const headers = this.getHeaders();
    return this.http.get<any[]>(`${this.apiUrl}/reportedComments`, { headers });
  }
}
